ext/dom/token_list.c

```diff
@@ -0,0 +1,729 @@
+	} else {
+		intern->cached_string = NULL;
+	}
+
+	if (should_free) {
+		xmlFree(value);
+	}
+}
+
+static void dom_token_list_ensure_set_up_to_date(dom_token_list_object *intern)
+{
+	bool should_free;
+	const xmlAttr *attr = dom_token_list_get_attr(intern);
+	xmlChar *value = dom_token_list_get_class_value(attr, &should_free);
+
+	/* xmlStrEqual will automatically handle equality rules of NULL vs "" (etc.) correctly. */
+	if (!xmlStrEqual(value, (const xmlChar *) intern->cached_string)) {
+		php_libxml_invalidate_cache_tag(&intern->cache_tag);
+		efree(intern->cached_string);
+		HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+		zend_hash_destroy(token_set);
+		zend_hash_init(token_set, 0, NULL, NULL, false);
+		dom_token_list_update_set(intern, token_set);
+	}
+
+	if (should_free) {
+		xmlFree(value);
+	}
+}
+
+void dom_token_list_ctor(dom_token_list_object *intern, dom_object *element_obj)
+{
+	php_libxml_node_ptr *ptr = element_obj->ptr;
+	ptr->refcount++;
+	intern->dom.ptr = ptr;
+	element_obj->document->refcount++;
+	intern->dom.document = element_obj->document;
+
+	intern->cache_tag.modification_nr = 0;
+
+	HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+	zend_hash_init(token_set, 0, NULL, NULL, false);
+
+	dom_token_list_update_set(intern, token_set);
+}
+
+void dom_token_list_free_obj(zend_object *object)
+{
+	dom_token_list_object *intern = php_dom_token_list_from_obj(object);
+
+	zend_object_std_dtor(object);
+
+	if (EXPECTED(intern->dom.ptr != NULL)) { /* Object initialized? */
+		xmlNodePtr node = dom_token_list_get_element(intern);
+		if (php_libxml_decrement_node_ptr_ref(intern->dom.ptr) == 0) {
+			php_libxml_node_free_resource(node);
+		}
+		php_libxml_decrement_doc_ref((php_libxml_node_object *) &intern->dom);
+		HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+		zend_hash_destroy(token_set);
+		efree(intern->cached_string);
+	}
+}
+
+static bool dom_token_list_item_exists(dom_token_list_object *token_list, zend_long index)
+{
+	dom_token_list_ensure_set_up_to_date(token_list);
+
+	HashTable *token_set = TOKEN_LIST_GET_SET(token_list);
+	return index >= 0 && index < zend_hash_num_elements(token_set);
+}
+
+static void dom_token_list_item_read(dom_token_list_object *token_list, zval *retval, zend_long index)
+{
+	dom_token_list_ensure_set_up_to_date(token_list);
+
+	HashTable *token_set = TOKEN_LIST_GET_SET(token_list);
+	if (index >= 0 && index < zend_hash_num_elements(token_set)) {
+		HashPosition position;
+		zend_hash_internal_pointer_reset_ex(token_set, &position);
+		while (index > 0) {
+			zend_hash_move_forward_ex(token_set, &position);
+			index--;
+		}
+		zend_string *str_index;
+		zend_hash_get_current_key_ex(token_set, &str_index, NULL, &position);
+		ZVAL_STR_COPY(retval, str_index);
+	} else {
+		/* Not an out of bounds ValueError, but NULL, as according to spec.
+		 * This was a design choice to allow for constructs like `item(x) ?? ...` */
+		ZVAL_NULL(retval);
```

Comment:
https://dom.spec.whatwg.org/#interface-domtokenlist states DOMTokenList implements `iterable<DOMString>;`.

From https://webidl.spec.whatwg.org/#idl-iterable:

> If a single type parameter is given, then the interface has a value iterator and provides values of the specified type. 

This applies in this case, reading further below:

> A value iterator must only be declared on an interface that supports indexed properties. The value-type of the value iterator must be the same as the type returned by the indexed property getter. A value iterator is implicitly defined to iterate over the object’s indexed properties.

So we must support indexed properties.

From https://webidl.spec.whatwg.org/#dfn-support-indexed-properties:

> An interface that defines an indexed property getter is said to support indexed properties.

And indexed property getter is defined here: https://webidl.spec.whatwg.org/#dfn-indexed-property-getter

Down below in their note they give an example of how an out of bounds access evaluates to undefined, which would map to NULL for us. This would also be consistent with how out-of-bounds array accesses in PHP result in NULL.

I'll try to write everything here down concisely in the comments.